File: server/modules/settings/router.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { requireSession } from "../auth/middleware.js";
import type { Session } from "../auth/session.js";
import type { LanManager } from "./lan.js";
import { getLanIP } from "./network.js";
import {
  createApprovedResumePrefixSchema,
  lanToggleBodySchema,
  patchTerminalSettingsSchema,
} from "../../schemas/api.js";
import type { SettingsStore } from "./store.js";
import {
  addApprovedResumePrefix,
  getTerminalSettings,
  listApprovedResumePrefixes,
  patchTerminalSettings,
  revokeApprovedResumePrefix,
  type PatchTerminalSettings,
} from "./terminal-settings.js";

export function createSettingsRouter(
  session: Session,
  lan: LanManager,
  onLanToggle: (enabled: boolean) => Promise<void>,
  settingsStore: SettingsStore,
) {
  return new Hono()
    .use("*", requireSession(session))
    .get("/lan", (c) => c.json(lan.getPublicState()))
    .post("/lan", zValidator("json", lanToggleBodySchema), async (c) => {
      const { enabled } = c.req.valid("json");

      if (enabled) {
        const ip = getLanIP();
        if (ip === "127.0.0.1") {
          return c.json({ error: "No network interface found" }, 503);
        }
      }

      try {
        await onLanToggle(enabled);
      } catch {
        return c.json({ error: "Failed to change network mode" }, 500);
      }

      return c.json(lan.getPublicState());
    })
    .post("/lan/refresh-invite", (c) => {
      if (lan.mode !== "lan") {
        return c.json({ error: "LAN is not enabled" }, 400);
      }
      lan.refreshInvite();
      return c.json(lan.getPublicState());
    })
    .get("/terminal", async (c) => {
      const settings = await getTerminalSettings(settingsStore);
      return c.json(settings);
    })
    .patch("/terminal", zValidator("json", patchTerminalSettingsSchema), async (c) => {
      const patch = c.req.valid("json") as PatchTerminalSettings;
      const settings = await patchTerminalSettings(settingsStore, patch);
      return c.json(settings);
    })
    .get("/terminal/resume-commands", async (c) => {
      const approvedPrefixes = await listApprovedResumePrefixes(settingsStore);
      return c.json({ approvedPrefixes });
    })
    .post(
      "/terminal/resume-commands",
      zValidator("json", createApprovedResumePrefixSchema),
      async (c) => {
        const body = c.req.valid("json");
        const entry = await addApprovedResumePrefix(settingsStore, body);
        return c.json({ prefix: entry }, 201);
      },
    )
    .delete("/terminal/resume-commands/:id", async (c) => {
      const ok = await revokeApprovedResumePrefix(settingsStore, c.req.param("id"));
      if (!ok) return c.json({ error: "Not found" }, 404);
      return c.json({ ok: true as const });
    });
}

File: server-go/internal/settings/router.go
package settings

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"
	"github.com/blaisetiong/workbench-cli/server-go/internal/auth"
	"github.com/blaisetiong/workbench-cli/server-go/internal/config"
)

// NetworkProvider supplies local URL and port for network settings.
type NetworkProvider interface {
	GetLocalURL() string
	GetURLScheme() string
	GetLocalHost() string
	Port() int
}

// NetworkSettings matches networkSettingsSchema.
type NetworkSettings struct {
	Host           string `json:"host"`
	Port           int    `json:"port"`
	ProdPort       int    `json:"prodPort"`
	NonProdPort    int    `json:"nonProdPort"`
	LocalURL       string `json:"localUrl"`
	Scheme         string `json:"scheme"`
	HostsFileLine  string `json:"hostsFileLine"`
	PendingRestart bool   `json:"pendingRestart"`
}

func buildNetworkSettings(net NetworkProvider) NetworkSettings {
	saved := config.LoadNetworkConfig()
	running := config.NetworkConfig{Host: net.GetLocalHost(), Port: net.Port()}
	pendingRestart := saved.Host != running.Host || saved.Port != running.Port
	return NetworkSettings{
		Host:           saved.Host,
		Port:           saved.Port,
		ProdPort:       saved.Port,
		NonProdPort:    config.NonProdPort(saved.Port),
		LocalURL:       net.GetLocalURL(),
		Scheme:         net.GetURLScheme(),
		HostsFileLine:  fmt.Sprintf("127.0.0.1 %s", saved.Host),
		PendingRestart: pendingRestart,
	}
}

func jsonResp(w http.ResponseWriter, v any, code int) {
	w.Header().Set("Content-Type", "application/json")
	if code != http.StatusOK {
		w.WriteHeader(code)
	}
	_ = json.NewEncoder(w).Encode(v)
}

func RegisterRoutes(r chi.Router, session *auth.Session, store Store, net NetworkProvider) {
	r.Use(auth.RequireSession(session))

	r.Get("/network", func(w http.ResponseWriter, r *http.Request) {
		jsonResp(w, buildNetworkSettings(net), http.StatusOK)
	})
	r.Patch("/network", func(w http.ResponseWriter, r *http.Request) {
		var patch struct {
			Host *string `json:"host,omitempty"`
			Port *int    `json:"port,omitempty"`
		}
		if err := json.NewDecoder(r.Body).Decode(&patch); err != nil {
			jsonResp(w, map[string]string{"error": "Bad request"}, http.StatusBadRequest)
			return
		}
		if patch.Host == nil && patch.Port == nil {
			jsonResp(w, map[string]string{"error": "No changes provided"}, http.StatusBadRequest)
			return
		}
		saved := config.LoadNetworkConfig()
		if patch.Host != nil {
			h := strings.TrimSpace(*patch.Host)
			if h == "" || len(h) > 253 {
				jsonResp(w, map[string]string{"error": "Invalid host"}, http.StatusBadRequest)
				return
			}
			saved.Host = h
		}
		if patch.Port != nil {
			if *patch.Port < 1 || *patch.Port > 65535 {
				jsonResp(w, map[string]string{"error": "Invalid port"}, http.StatusBadRequest)
				return
			}
			saved.Port = *patch.Port
		}
		if err := config.SaveNetworkConfig(saved); err != nil {
			jsonResp(w, map[string]string{"error": err.Error()}, http.StatusInternalServerError)
			return
		}
		jsonResp(w, buildNetworkSettings(net), http.StatusOK)
	})

	r.Get("/terminal", func(w http.ResponseWriter, r *http.Request) {
		jsonResp(w, GetTerminalSettings(store), http.StatusOK)
	})
	r.Patch("/terminal", func(w http.ResponseWriter, r *http.Request) {
		var patch PatchTerminalSettings
		if err := json.NewDecoder(r.Body).Decode(&patch); err != nil {
			jsonResp(w, map[string]string{"error": "Bad request"}, http.StatusBadRequest)
			return
		}
		jsonResp(w, PatchTerminalSettingsFn(store, patch), http.StatusOK)
	})

	r.Get("/terminal/resume-commands", func(w http.ResponseWriter, r *http.Request) {
		jsonResp(w, map[string]any{"approvedPrefixes": ListApprovedResumePrefixes(store)}, http.StatusOK)
	})
	r.Post("/terminal/resume-commands", func(w http.ResponseWriter, r *http.Request) {
		var input CreateApprovedResumePrefix
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil || strings.TrimSpace(input.Prefix) == "" {
			jsonResp(w, map[string]string{"error": "prefix is required"}, http.StatusBadRequest)
			return
		}
		input.Prefix = strings.TrimSpace(input.Prefix)
		entry := AddApprovedResumePrefix(store, input)
		jsonResp(w, map[string]any{"prefix": entry}, http.StatusCreated)
	})
	r.Delete("/terminal/resume-commands/{id}", func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		if !RevokeApprovedResumePrefix(store, id) {
			jsonResp(w, map[string]string{"error": "Not found"}, http.StatusNotFound)
			return
		}
		jsonResp(w, map[string]bool{"ok": true}, http.StatusOK)
	})
}
